### ISSUE TITLE
how to implement fallback

### ISSUE THREAD

**Issue by Primtek** [RasaHQ/rasa_core#700]
**Rasa Core version**:
0.9.7

**Python version**: 
2.7.5

**Operating system** (windows, osx, ...):
centOS 7

**Issue**:
I m new, how to implement fallback? 
I found some codes, but I don't know where to put those codes .. in bot.py? or policy.py?
it tried it but it does not work :(

bot.py

class ActionFallback(Action):
    def name(self):
        return "action_fallback"

    def run(self, dispatcher, tracker, domain):
        dispatcher.utter_message("Sorry, didn't get that. Try again.")
        return [AllSlotsReset()]

def train_dialogue(domain_file="domain.yml",
                   model_path="models/dialogue",
                   training_data_file="data/stories.md"):

    fallback = FallbackPolicy(fallback_action_name="utter_unknown",
                          core_threshold=0.2,
                          nlu_threshold=0.2)
    agent = Agent(domain_file, policies=[
        AugmentedMemoizationPolicy(max_history=max_history),
        KerasPolicy(),fallback])
    training_data = agent.load_data(training_data_file="data/stories.md")

    #agent = Agent(domain_file="domain.yml",
    #              policies=[MemoizationPolicy(max_history=3),
    #                        RestaurantPolicy()])

    #training_data = agent.load_data(training_data_file="data/stories.md")

    agent.train(
            training_data,
            epochs=400,
            batch_size=100,
            validation_split=0.2
    )

    agent.persist(model_path)
    return agent

very appreciate if someone can guide me for this..
thanks a lot in advance

**Content of domain file** (if used & relevant):
```yaml
domain.yml
templates:
   ....
  utter_unknown:
    - "Sorry, I don't understand"
```
stories.md
.....
## fallback
  - utter_unknown

**Comment by Primtek** [comment]
I also followed: https://core.rasa.com/patterns.html#fallback-default-actions
no luck :(

**Comment by amn41** [comment]
what doesn't work exactly? could you please share the output if you run with `rasa_core.run --debug` ?

**Comment by Primtek** [comment]
I put fall back as follow:

def train_dialogue(domain_file="domain.yml",
                   model_path="models/dialogue",
                   training_data_file="data/stories.md"):


    from rasa_core.policies.fallback import FallbackPolicy
    from rasa_core.policies.keras_policy import KerasPolicy
    from rasa_core.agent import Agent

    fallback = FallbackPolicy(fallback_action_name="utter_default",
                          core_threshold=0.3,
                          nlu_threshold=0.3)

    agent = Agent("domain.yml",
               policies=[KerasPolicy(), fallback])

    training_data = agent.load_data(training_data_file)

    agent.train(
            training_data,
            epochs=400,
            batch_size=100,
            validation_split=0.2
    )

    agent.persist(model_path)
    return agent


python -m rasa_core.run --debug -d models/dialogue -u models/nlu/default/current/

Rasa process starting
/usr/lib64/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters

Added 'nlp_spacy' to component cache. Key 'nlp_spacy-en'.
Finished loading agent, starting input channel & server.
Creating a new tracker for id 'default'.
/usr/lib64/python2.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:

Received user message 'hey' with intent '{u'confidence': 0.9024305568420068, u'name': u'greet'}' and entities '[]'
Logged UserUtterance - tracker now has 2 events
Current slot values:
        info: None
        cuisine: None
        people: None
        matches: None
        price: None
        location: None
Predicted next action using policy_0_KerasPolicy
Predicted next action 'action_listen' with prob 0.75.
Action 'action_listen' ended with events '[]'
Current topic: None
Recreating tracker for id 'default'
/usr/lib64/python2.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:

Received user message 'blahblahxxxx' with intent '{u'confidence': 0.8423502315853164, u'name': u'greet'}' and entities '[{u'extractor': u'ner_crf', u'confidence': 0.5257591050924016, u'end': 12, u'value': u'blahblahxxxx', u'entity': 'cuisine', u'start': 0}]'
Logged UserUtterance - tracker now has 5 events
Current slot values:
        info: None
        cuisine: blahblahxxxx
        people: None
        matches: None
        price: None
        location: None
Predicted next action using policy_0_KerasPolicy
Predicted next action 'utter_on_it' with prob 0.99.
Bot utterance 'BotUttered(text: I'm on it, data: null)'
Action 'utter_on_it' ended with events '[]'
Predicted next action using policy_0_KerasPolicy
Predicted next action 'utter_ask_location' with prob 0.86.
Bot utterance 'BotUttered(text: where?, data: null)'
Action 'utter_ask_location' ended with events '[]'
Predicted next action using policy_0_KerasPolicy
Predicted next action 'utter_on_it' with prob 0.99.
Bot utterance 'BotUttered(text: I'm on it, data: null)'
Action 'utter_on_it' ended with events '[]'
Predicted next action using policy_0_KerasPolicy
Predicted next action 'utter_ask_location' with prob 0.86.
Bot utterance 'BotUttered(text: where?, data: null)'
Action 'utter_ask_location' ended with events '[]'
Predicted next action using policy_0_KerasPolicy
Predicted next action 'action_listen' with prob 1.00.
Action 'action_listen' ended with events '[]'
Current topic: None
Recreating tracker for id 'default'
/usr/lib64/python2.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:

Received user message 'hhhxxxxaaathis should be unknown' with intent '{u'confidence': 0.7926958313618006, u'name': u'inform'}' and entities '[]'
Logged UserUtterance - tracker now has 11 events
Current slot values:
        info: None
        cuisine: blahblahxxxx
        people: None
        matches: None
        price: None
        location: None
Predicted next action using policy_0_KerasPolicy
Predicted next action 'utter_ask_numpeople' with prob 0.96.
Bot utterance 'BotUttered(text: for how many people?, data: null)'
Action 'utter_ask_numpeople' ended with events '[]'
Predicted next action using policy_0_KerasPolicy
Predicted next action 'action_listen' with prob 1.00.
Action 'action_listen' ended with events '[]'
Current topic: None

**Comment by akelad** [comment]
I think this was answered in a different issue now, so i'll close this one. You need to re-train the Core model before the fallback policy will work

**Comment by GlorianY** [comment]
Hi @akelad. Could you give the link to the issue which answers this question? Thanks!
